<a href="https://colab.research.google.com/github/Adizcool/Job_Promotion_Prediction/blob/main/Job_Promotion_Prediction_Data_Preprocessing_2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**1) Install/ Import the required Python Packages/ Libraries**

In [1]:
#Import required python packages
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sb
from sklearn import preprocessing
%matplotlib inline

**2) Mounting the Google Drive**

In [2]:
# Mount the Google Drive
from google.colab import drive
drive.mount('/content/gdrive')

Mounted at /content/gdrive


**3) Read the Data file and check**

In [3]:
# Read the Diabetes Data from .csv file and check the data shape (number of Rows and Columns)
#df = pd.read_csv('gdrive/My Drive/Indi_ML_Training/Training-Sessions/MLP-Session25-05-Jun-2021/train_ctrUa4K.csv')
df = pd.read_csv("gdrive/My Drive/Datasets/HR Analysis/train_HR_Analytics-Preprocessed.csv")
print(df.shape)
df.head()

(5000, 20)


education  no_of_trainings  ...  recruitment_channel_sourcing  is_promoted
0   1.531419        -0.406222  ...                      1.182891            0
1  -0.664164        -0.406222  ...                     -0.845386            0
2  -0.664164        -0.406222  ...                      1.182891            0
3  -0.664164         1.420317  ...                     -0.845386            0
4  -0.664164        -0.406222  ...                     -0.845386            0

[5 rows x 20 columns]

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5000 entries, 0 to 4999
Data columns (total 20 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   education                     5000 non-null   float64
 1   no_of_trainings               5000 non-null   float64
 2   age                           5000 non-null   float64
 3   previous_year_rating          5000 non-null   float64
 4   length_of_service             5000 non-null   float64
 5   KPIs_met>80%                  5000 non-null   float64
 6   awards_won?                   5000 non-null   float64
 7   avg_training_score            5000 non-null   float64
 8   department_Finance            5000 non-null   float64
 9   department_HR                 5000 non-null   float64
 10  department_Legal              5000 non-null   float64
 11  department_Operations         5000 non-null   float64
 12  department_Procurement        5000 non-null   float64
 13  dep

##**VII) Check the Test accuracy using appropriate algorithm and Holdout Method.**

##**Step-5: Slice X and y Values**

In [5]:
X = df.drop(['is_promoted'], axis = 1)
Y = df['is_promoted']
X.head()

education  ...  recruitment_channel_sourcing
0   1.531419  ...                      1.182891
1  -0.664164  ...                     -0.845386
2  -0.664164  ...                      1.182891
3  -0.664164  ...                     -0.845386
4  -0.664164  ...                     -0.845386

[5 rows x 19 columns]

In [6]:
Y.head()

0    0
1    0
2    0
3    0
4    0
Name: is_promoted, dtype: int64

##**Step-6: Execute Train-Test-Split Command and Verify**

In [7]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, Y, test_size = 0.2, random_state = 66)

In [8]:
print(X_train.shape)
print(y_train.shape)
print(X_test.shape)
print(y_test.shape)

(4000, 19)
(4000,)
(1000, 19)
(1000,)


##**Step-7: Learn the Data and Predict the dependent Variable values for the "X_test"data using "LogisticRegression()" algorithm**

In [9]:
from sklearn.linear_model import LogisticRegression
#create an instance and fit the model 
logmodel = LogisticRegression()
logmodel.fit(X_train, y_train)

LogisticRegression()

In [10]:
y_pred = logmodel.predict(X_test)
y_pred

array([0, 0, 1, 1, 1, 0, 0, 1, 0, 1, 0, 0, 1, 0, 0, 1, 1, 0, 1, 1, 0, 0,
       0, 1, 0, 0, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 1, 1, 1, 0,
       1, 1, 0, 1, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0,
       1, 0, 0, 0, 0, 1, 1, 1, 0, 0, 1, 1, 0, 0, 1, 1, 1, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 1, 0, 1, 1, 0, 0, 1, 0, 1, 0,
       1, 0, 1, 1, 0, 0, 0, 0, 1, 1, 0, 1, 1, 0, 1, 0, 0, 0, 0, 1, 1, 0,
       0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 1, 1, 1, 0, 0, 1, 1, 1, 0, 0, 0, 1,
       0, 1, 1, 1, 1, 0, 0, 1, 0, 1, 0, 1, 1, 0, 0, 0, 1, 1, 1, 1, 1, 1,
       0, 1, 1, 0, 1, 0, 0, 0, 1, 1, 1, 1, 1, 0, 1, 0, 0, 0, 0, 0, 1, 0,
       0, 1, 1, 0, 0, 0, 1, 1, 0, 1, 1, 0, 0, 1, 1, 0, 1, 0, 0, 1, 1, 1,
       1, 0, 0, 0, 1, 1, 1, 0, 0, 1, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0,
       0, 0, 0, 1, 0, 1, 0, 0, 0, 1, 0, 1, 1, 1, 1, 0, 1, 0, 1, 0, 1, 1,
       0, 1, 1, 1, 0, 0, 1, 1, 1, 1, 0, 0, 0, 0, 1, 1, 0, 1, 0, 0, 1, 1,
       0, 1, 0, 0, 1, 1, 1, 1, 0, 1, 0, 1, 1, 0, 1,

##**Step-8: Calculate the Accuracy of the Model**

In [11]:
accuracy_lr = logmodel.score(X_test, y_test)
print("Accuracy of Logistic Regression on test set:",accuracy_lr)

Accuracy of Logistic Regression on test set: 0.768


##**Step-9: Display the Confusion Matrix and Classification Report of the Model**

In [12]:
from sklearn.metrics import classification_report, confusion_matrix  
print(confusion_matrix(y_test,y_pred))  
print(classification_report(y_test,y_pred))  

[[412 138]
 [ 94 356]]
              precision    recall  f1-score   support

           0       0.81      0.75      0.78       550
           1       0.72      0.79      0.75       450

    accuracy                           0.77      1000
   macro avg       0.77      0.77      0.77      1000
weighted avg       0.77      0.77      0.77      1000



##**IX) Check for Multi-collinearity and find out the new accuracy as required**

###**Checking for MultiCollinearity**

VIF starts at 1 and has no upper limit

VIF = 1, no correlation between the independent variable and the other variables

VIF exceeding 5 or 10 indicates high multicollinearity between this independent variable and the others

In [13]:
# Import library for VIF
from statsmodels.stats.outliers_influence import variance_inflation_factor

def calc_vif(X):

    # Calculating VIF
    vif = pd.DataFrame()
    vif["variables"] = X.columns
    vif["VIF"] = [variance_inflation_factor(X.values, i) for i in range(X.shape[1])]

    return(vif)

/usr/local/lib/python3.7/dist-packages/statsmodels/tools/_testing.py:19: FutureWarning: pandas.util.testing is deprecated. Use the functions in the public API at pandas.testing instead.
  import pandas.util.testing as tm


In [14]:
calc_vif(X)

variables       VIF
0                      education  1.203711
1                no_of_trainings  1.038746
2                            age  2.068588
3           previous_year_rating  1.140835
4              length_of_service  1.870183
5                   KPIs_met>80%  1.111005
6                    awards_won?  1.095883
7             avg_training_score  3.247848
8             department_Finance  1.619688
9                  department_HR  2.018490
10              department_Legal  1.271776
11         department_Operations  3.831265
12        department_Procurement  2.550431
13                department_R&D  1.152490
14  department_Sales & Marketing  5.813208
15          deartment_Technology  2.196660
16                      gender_m  1.129651
17  recruitment_channel_referred  1.035444
18  recruitment_channel_sourcing  1.021894

In [15]:
X.shape

(5000, 19)

In [16]:
X1= X.drop(['department_Sales & Marketing'], axis=1, inplace=False)

In [17]:
X1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5000 entries, 0 to 4999
Data columns (total 18 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   education                     5000 non-null   float64
 1   no_of_trainings               5000 non-null   float64
 2   age                           5000 non-null   float64
 3   previous_year_rating          5000 non-null   float64
 4   length_of_service             5000 non-null   float64
 5   KPIs_met>80%                  5000 non-null   float64
 6   awards_won?                   5000 non-null   float64
 7   avg_training_score            5000 non-null   float64
 8   department_Finance            5000 non-null   float64
 9   department_HR                 5000 non-null   float64
 10  department_Legal              5000 non-null   float64
 11  department_Operations         5000 non-null   float64
 12  department_Procurement        5000 non-null   float64
 13  dep

In [18]:
calc_vif(X1)

variables       VIF
0                      education  1.200882
1                no_of_trainings  1.032601
2                            age  2.067615
3           previous_year_rating  1.138386
4              length_of_service  1.870152
5                   KPIs_met>80%  1.106878
6                    awards_won?  1.053079
7             avg_training_score  1.547854
8             department_Finance  1.070916
9                  department_HR  1.115773
10              department_Legal  1.030201
11         department_Operations  1.323697
12        department_Procurement  1.410920
13                department_R&D  1.096222
14          deartment_Technology  1.578265
15                      gender_m  1.129270
16  recruitment_channel_referred  1.035041
17  recruitment_channel_sourcing  1.021743

In [19]:
from sklearn.model_selection import train_test_split
X1_train, X1_test, Y1_train, Y1_test = train_test_split(X1, Y, test_size = 0.2, random_state = 66)

In [20]:
from sklearn.linear_model import LogisticRegression
#create an instance and fit the model 
logmodel2 = LogisticRegression()
logmodel2.fit(X1_train, Y1_train)

LogisticRegression()

In [21]:
#predictions
predictions1 = logmodel2.predict(X1_test)

In [22]:
print(confusion_matrix(Y1_test, predictions1))
print(classification_report(Y1_test,predictions1))

[[410 140]
 [151 299]]
              precision    recall  f1-score   support

           0       0.73      0.75      0.74       550
           1       0.68      0.66      0.67       450

    accuracy                           0.71      1000
   macro avg       0.71      0.70      0.71      1000
weighted avg       0.71      0.71      0.71      1000



**Observation: After removal of the column " department_Sales & Marketing" the model gives the accuracy of 71%**

**Decision: We will not remove the column "department_Sales & Marketing"**

##**X) Check for Outliers and handle them accordingly**

###**Outliers Detection**

In [23]:
df1=df

In [24]:
df1.head()

education  no_of_trainings  ...  recruitment_channel_sourcing  is_promoted
0   1.531419        -0.406222  ...                      1.182891            0
1  -0.664164        -0.406222  ...                     -0.845386            0
2  -0.664164        -0.406222  ...                      1.182891            0
3  -0.664164         1.420317  ...                     -0.845386            0
4  -0.664164        -0.406222  ...                     -0.845386            0

[5 rows x 20 columns]

In [25]:
from scipy import stats
import numpy as np
z = np.abs(stats.zscore(df1))
print(z[:9])

[[1.53141901 0.40622225 0.04392902 1.21087568 0.49798543 1.01369375
  0.24768082 1.2260367  0.20518498 0.20252415 0.12340351 0.51340386
  0.41314165 0.13150665 1.59643855 0.41909513 1.510958   0.16338312
  1.18289148 0.91076419]
 [0.6641638  0.40622225 0.6228746  1.21087568 0.43605833 0.98649123
  0.24768082 0.46645328 0.20518498 0.20252415 0.12340351 1.94778434
  0.41314165 0.13150665 0.6263943  0.41909513 0.66183176 0.16338312
  0.84538609 0.91076419]
 [0.6641638  0.40622225 0.0894317  0.4876454  0.26447449 0.98649123
  0.24768082 1.15698366 0.20518498 0.20252415 0.12340351 0.51340386
  0.41314165 0.13150665 1.59643855 0.41909513 0.66183176 0.16338312
  1.18289148 0.91076419]
 [0.6641638  1.42031664 0.57737192 2.18616649 0.96500732 0.98649123
  0.24768082 1.15698366 0.20518498 0.20252415 0.12340351 0.51340386
  0.41314165 0.13150665 1.59643855 0.41909513 0.66183176 0.16338312
  0.84538609 0.91076419]
 [0.6641638  0.40622225 1.37753627 0.4876454  0.90308021 0.98649123
  0.24768082 0.4

In [26]:
z.shape

(5000, 20)

In [27]:
df2 = df1

In [28]:
threshold = 3
print(np.where(z > 3))

(array([  14,   22,   24, ..., 4996, 4996, 4999]), array([13,  8,  8, ...,  1,  6,  1]))


In [29]:
df2 = df2[(z < 3).all(axis=1)]

In [30]:
df2.shape

(3822, 20)

In [31]:
X3 = df.drop(['is_promoted'], axis = 1)
Y3 = df['is_promoted']
X3.head()

education  ...  recruitment_channel_sourcing
0   1.531419  ...                      1.182891
1  -0.664164  ...                     -0.845386
2  -0.664164  ...                      1.182891
3  -0.664164  ...                     -0.845386
4  -0.664164  ...                     -0.845386

[5 rows x 19 columns]

In [32]:
columnNames = ['education','no_of_trainings', 'age', 'previous_year_rating', 'length_of_service', 'KPIs_met>80%', 'awards_won?', 'avg_training_score', 'department_Finance',
               'department_HR', 'department_Legal', 'department_Operations', 'department_Procurement', 'department_R&D', 'department_Sales & Marketing', 'deartment_Technology',
               'gender_m', 'recruitment_channel_referred', 'recruitment_channel_sourcing']

In [33]:
std_scaler_object = preprocessing.StandardScaler()
X3A = std_scaler_object.fit_transform(X3)
X3A = pd.DataFrame(X3A , columns = columnNames)
X3A.head()

education  ...  recruitment_channel_sourcing
0   1.531419  ...                      1.182891
1  -0.664164  ...                     -0.845386
2  -0.664164  ...                      1.182891
3  -0.664164  ...                     -0.845386
4  -0.664164  ...                     -0.845386

[5 rows x 19 columns]

In [34]:
from sklearn.model_selection import train_test_split
X3_train, X3_test, Y3_train, Y3_test = train_test_split(X3A, Y3, test_size = 0.2, random_state = 66)

In [35]:
from sklearn.linear_model import LogisticRegression
#create an instance and fit the model 
logmodel2 = LogisticRegression()
logmodel2.fit(X3_train, Y3_train)

LogisticRegression()

In [36]:
#predictions
predictions4 = logmodel2.predict(X3_test)

In [37]:
print(confusion_matrix(Y3_test, predictions4))
print(classification_report(Y3_test,predictions4))

[[412 138]
 [ 94 356]]
              precision    recall  f1-score   support

           0       0.81      0.75      0.78       550
           1       0.72      0.79      0.75       450

    accuracy                           0.77      1000
   macro avg       0.77      0.77      0.77      1000
weighted avg       0.77      0.77      0.77      1000



**Observation: After Outliers removal we get the accuracy of 77%**

**Decision: We will not remove the Outliers Data.**

##**XI) Check for Feature Selection/ Extraction and verify the accuracy**

###**Using Correlation Coefficient Method**

In [38]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5000 entries, 0 to 4999
Data columns (total 20 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   education                     5000 non-null   float64
 1   no_of_trainings               5000 non-null   float64
 2   age                           5000 non-null   float64
 3   previous_year_rating          5000 non-null   float64
 4   length_of_service             5000 non-null   float64
 5   KPIs_met>80%                  5000 non-null   float64
 6   awards_won?                   5000 non-null   float64
 7   avg_training_score            5000 non-null   float64
 8   department_Finance            5000 non-null   float64
 9   department_HR                 5000 non-null   float64
 10  department_Legal              5000 non-null   float64
 11  department_Operations         5000 non-null   float64
 12  department_Procurement        5000 non-null   float64
 13  dep

In [39]:
cor=df.corr()
cor

education  ...  is_promoted
education                      1.000000  ...     0.050483
no_of_trainings               -0.051637  ...    -0.049298
age                            0.383352  ...    -0.037035
previous_year_rating           0.035436  ...     0.287039
length_of_service              0.266965  ...    -0.019550
KPIs_met>80%                   0.032528  ...     0.388057
awards_won?                    0.018963  ...     0.216859
avg_training_score             0.034312  ...     0.271689
department_Finance            -0.024766  ...    -0.001197
department_HR                  0.043806  ...    -0.031636
department_Legal              -0.036803  ...    -0.023153
department_Operations          0.036197  ...     0.020867
department_Procurement         0.109469  ...     0.000537
department_R&D                 0.034945  ...    -0.023430
department_Sales & Marketing  -0.062686  ...    -0.040937
deartment_Technology          -0.007967  ...     0.056699
gender_m                      -0.069929  ...    -0.014378
recruitment_channel_referred  -0.034010  ...     0.038015
recruitment_channel_sourcing   0.001416  ...    -0.012945
is_promoted                    0.050483  ...     1.000000

[20 rows x 20 columns]

In [40]:
threshold=0.3
a = abs(cor['is_promoted'])
a

education                       0.050483
no_of_trainings                 0.049298
age                             0.037035
previous_year_rating            0.287039
length_of_service               0.019550
KPIs_met>80%                    0.388057
awards_won?                     0.216859
avg_training_score              0.271689
department_Finance              0.001197
department_HR                   0.031636
department_Legal                0.023153
department_Operations           0.020867
department_Procurement          0.000537
department_R&D                  0.023430
department_Sales & Marketing    0.040937
deartment_Technology            0.056699
gender_m                        0.014378
recruitment_channel_referred    0.038015
recruitment_channel_sourcing    0.012945
is_promoted                     1.000000
Name: is_promoted, dtype: float64

In [41]:
result = a[a>0.3]
result

KPIs_met>80%    0.388057
is_promoted     1.000000
Name: is_promoted, dtype: float64

In [42]:
df4=df.drop(['education','no_of_trainings', 'age', 'previous_year_rating', 'length_of_service', 'awards_won?', 'avg_training_score', 'department_Finance',
               'department_HR', 'department_Legal', 'department_Operations', 'department_Procurement', 'department_R&D', 'department_Sales & Marketing', 'deartment_Technology',
               'gender_m', 'recruitment_channel_referred', 'recruitment_channel_sourcing'], axis=1)

In [43]:
df4.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5000 entries, 0 to 4999
Data columns (total 2 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   KPIs_met>80%  5000 non-null   float64
 1   is_promoted   5000 non-null   int64  
dtypes: float64(1), int64(1)
memory usage: 78.2 KB


In [44]:
X5 = df4.drop(['is_promoted'], axis = 1)
Y5 = df4['is_promoted']
X5.head()

KPIs_met>80%
0      1.013694
1     -0.986491
2     -0.986491
3     -0.986491
4     -0.986491

In [45]:
Y5.head()

0    0
1    0
2    0
3    0
4    0
Name: is_promoted, dtype: int64

In [46]:
from sklearn.model_selection import train_test_split
X5_train, X5_test, Y5_train, Y5_test = train_test_split(X5, Y5, test_size = 0.2, random_state = 66)

In [47]:
print(X5_train.shape)
print(Y5_train.shape)
print(X5_test.shape)
print(Y5_test.shape)

(4000, 1)
(4000,)
(1000, 1)
(1000,)


In [48]:
from sklearn.linear_model import LogisticRegression
#create an instance and fit the model 
logmodel2 = LogisticRegression()
logmodel2.fit(X5_train, Y5_train)

LogisticRegression()

In [49]:
#predictions
predictions5 = logmodel2.predict(X5_test)

In [50]:
print(confusion_matrix(Y5_test, predictions5))
print(classification_report(Y5_test,predictions5))

[[369 181]
 [142 308]]
              precision    recall  f1-score   support

           0       0.72      0.67      0.70       550
           1       0.63      0.68      0.66       450

    accuracy                           0.68      1000
   macro avg       0.68      0.68      0.68      1000
weighted avg       0.68      0.68      0.68      1000



**Observation: By having only one independent variable column "KPIs_met>80%", we get the accuracy of 68%**

###**Decision: We cannot use "KPIs_met>80%" as the only Independent Variable Data**

###**Applying PCA**

In [51]:
from sklearn.decomposition import PCA

pca = PCA(n_components=2)
X_train6 = pca.fit_transform(X_train)
X_test6 = pca.transform(X_test)
print(X_train6.shape)
print(X_test6.shape)
print(X_train6[:5])
print(Y[:5])

(4000, 2)
(1000, 2)
[[ 1.22460944  0.13094297]
 [-1.84258267  1.34506597]
 [ 0.65647523 -1.38003196]
 [-1.83995575 -0.00465833]
 [ 0.32525361 -0.89904423]]
0    0
1    0
2    0
3    0
4    0
Name: is_promoted, dtype: int64


In [52]:
from sklearn.linear_model import LogisticRegression
#create an instance and fit the model 
logmodel2 = LogisticRegression()
logmodel2.fit(X_train6, y_train)

LogisticRegression()

In [53]:
#predictions
predictions6 = logmodel2.predict(X_test6)

In [54]:
from sklearn.metrics import classification_report, confusion_matrix  
print(confusion_matrix(y_test,predictions6))  
print(classification_report(y_test,predictions6))  

[[374 176]
 [226 224]]
              precision    recall  f1-score   support

           0       0.62      0.68      0.65       550
           1       0.56      0.50      0.53       450

    accuracy                           0.60      1000
   macro avg       0.59      0.59      0.59      1000
weighted avg       0.59      0.60      0.59      1000



**Observation: By using the PCA Method, we get the accuracy of 59%**

###**Decision: We will not use the PCA Method for reducing the features**

###**Using LDA**

In [55]:
from sklearn.discriminant_analysis import LinearDiscriminantAnalysis as LDA

lda = LDA(n_components=1)
X_train7 = lda.fit_transform(X_train, y_train)
X_test7 = lda.transform(X_test)
print(X_train7[:5])
print()
print(X_test7[:5])

[[ 3.2283724 ]
 [-0.903777  ]
 [-0.3328824 ]
 [-0.08878046]
 [-1.46928838]]

[[-1.2704975 ]
 [-1.79896896]
 [ 1.09019099]
 [ 3.70798063]
 [ 3.88926977]]


In [56]:
from sklearn.linear_model import LogisticRegression
#create an instance and fit the model 
logmodel2 = LogisticRegression()
logmodel2.fit(X_train7, y_train)

LogisticRegression()

In [57]:
#predictions
predictions7 = logmodel2.predict(X_test7)

In [58]:
from sklearn.metrics import classification_report, confusion_matrix  
print(confusion_matrix(y_test,predictions7))  
print(classification_report(y_test,predictions7))  

[[400 150]
 [105 345]]
              precision    recall  f1-score   support

           0       0.79      0.73      0.76       550
           1       0.70      0.77      0.73       450

    accuracy                           0.74      1000
   macro avg       0.74      0.75      0.74      1000
weighted avg       0.75      0.74      0.75      1000



**Observation: By using the LDA Method, we get the accuracy of 74%**

###**Decision: We will not use the LDA Method for reducing the features**